enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,666 @@
+
+### Implementation Details/Notes/Constraints
+
+#### High-level End-to-end Workflow
+
+This section outlines an end-to-end workflow for deploying a cluster
+with management-workload-partitioning enabled and how pods are
+correctly scheduled to run on the management CPU pool.
+
+1. User sits down at their computer.
+2. The user creates their `install-config.yaml`, including extra
+   values in the specifying the CPUs to include in the management CPU
+   set. (See below for details.)
+   * Does the user also need to explicitly enable the feature flag?
+3. The user runs the installer.
+4. The installer uses the management CPU pool settings to generate an
+   extra machine config manifest to configure CRI-O to process
+   management workloads in a special way.
+5. The installer creates a machine config manifest to write a
+   configuration file for kubelet. The file should only be readable by
+   the kubelet.
+6. The installer enables the feature set for the feature flag
+   controlling management workload partitioning.
+7. The kubelet starts up and finds the configuration file enabling the
+   new feature.
+8. The kubelet advertises `cpu.workload.openshift.io/management`
+   extended resources on the node based on the number of CPUs in the
+   host.
+9. The kubelet reads static pod definitions. It replaces the `cpus`
+   requests with `cpu.workload.openshift.io/management` requests of
+   the same value and adds the
+   `cpu.workload.openshift.io/{container-name}` annotations for CRI-O
+   with the same values.
+10. Something schedules a regular pod with the
+    `workload.openshift.io/management` label in a namespace with the
+    `workload.openshift.io/management` label.
+11. The admission hook modifies the pod, replacing the CPU requests
+    with `cpu.workload.openshift.io/management` requests and adding
+    the `cpu.workload.openshift.io/{container-name}` annotations for
+    CRI-O.
+12. The scheduler sees the new pod and finds available
+    `cpu.workload.openshift.io/management` resources on the node. The
+    scheduler places the pod on the node.
+13. Repeat steps 10-12 until all pods are running.
+14. Single-node deployment comes up with management components
+    constrained to subset of available CPUs
+
+#### Pod mutation
+
+The kubelet and API admission hook will change pods labeled with
+`workload.openshift.io/management` so the CPU requests are replaced
+with management CPU requests and an annotation is added with the same
+value.
+
+```yaml
+requests:
+  cpu:
+    400m
+```
+
+becomes
+
+```yaml
+requests:
+  cpu.workload.openshift.io/management: 400m
+```
+
+and
+
+```yaml
+annotations:
+  cpu.workload.openshift.io/container-name: 400m
+```
+
+
+#### CRI-O Changes
+
+*NOTE: This section needs to be updated to deal with the generalized
+CPU pools.*
+
+CRI-O will be updated to support a new configuration value that a user
+can specify, `mgmt_ctr_cpuset`.
+
+```ini
+[crio.runtime]
```

Comment:
We can make this work similar to how we specify runtime handlers today. e.g.
```
[crio.runtime.workloads.{workload-type}]
  cpu_set = "0-1"
```

For the management case:
```
[crio.runtime.workloads.management]
  cpu_set = "0-1"
```

In the future the configuration could be:
```
[crio.runtime.workloads.management]
  cpu_set = "0-1"

[crio.runtime.workloads.networking]
  cpu_set = "2-3"

```

cc: @haircommander

Reply:
This can also handle other settings besides cpu_set for workloads in the future.

Reply:
We can also add the `label` for activating this under the same setting and the `cpu_shares_annotation` to avoid any hard coding.

Reply:
> We can also add the `label` for activating this under the same setting and the `cpu_shares_annotation` to avoid any hard coding.

So, the installer would set up a machine config resource to configure CRI-O with a new workloads section, including the cpu_set and the name of the label (or annotation?) that will be used to enable use of those workload settings. That restricts all of the openshift-specific stuff to the installer, and let's CRI-O work from its config file. Is that right?

Reply:
Yes, right.

Reply:
I have tried to capture this in the latest iteration. See 340d2dd62c1e669d68a59c76c8cc989386514340